### PR TITLE
Install Fluxbox window manager in VNC images

### PR DESCRIPTION
**Pull request by bonigarcia**
This PR is related to #29. It basically installs [Fluxbox](http://fluxbox.org/) window manager in the VNC images of Selenoid. The normal images (i.e. without Selenoid) are nor affected, and the windows manager is only available for the VNC flavor of Selenoid images.

I have compared the the resulting images to the original ones, and as you can see the change has a little effect. I checked it for latest version of Chrome, Firefox, and Opera. I built these images using the following commands:

```
./automate_chrome.sh 64.0.3282.186-1 2.35 64.0
./automate_firefox.sh 58.0.2+build1-0ubuntu0.16.04.1 1.5.1 58.0 0.19.1
./automate_opera.sh 51.0.2830.40 2.33 51.0
```

The resulting images are as follows:

```
REPOSITORY          TAG                 IMAGE ID            CREATED             SIZE
selenoid/vnc        chrome_64.0         5b592a2c44f7        4 seconds ago       886MB
selenoid/vnc        firefox_58.0        01cfbc1552e3        9 seconds ago       915MB
selenoid/vnc        opera_51.0          6d155108ed9c        7 seconds ago       884MB
```

After installing Fluxbox in the images, they are as follows. Notice that the difference in terms of size is only **5 MB**:

```
REPOSITORY          TAG                 IMAGE ID            CREATED             SIZE
selenoid/vnc        chrome_64.0         9c60e605ffaf        25 seconds ago      891MB
selenoid/vnc        firefox_58.0        c0bee659410e        10 seconds ago      920MB
selenoid/vnc        opera_51.0          31c9e5d2730c        8 seconds ago       889MB
```

Of course the CPU/memory consumption is now a bit higher since the windows manager is running in the container, but it should not be big deal IMHO.

As main benefit of the change, now we now use the browsers exactly how they are out of Docker. This is quite useful in the case of using [noVNC](http://novnc.com/info.html) (which is my case), in which I interact with the browser using an HTML5 Canvas, and now I can maximize, open preferences, and so on (without a windows manager this was not possible). For example as follows:

image

Please consider to merge it to the branch master to be available in next VNC Selenoid images. Thanks in advance.

### CODE REVIEW

**Comment by vania-pooh** [comment]
@bonigarcia ok, discussing this with guys.

**Comment by bonigarcia** [comment]
@vania-pooh because you said in #29 that windows manager is not useful for automated browsers. To minimize the consumption, I just installed it in the VNC images, in which the windows manager can be useful. In any case, if you prefer I can change the PR installing Fluxbox in the base image.

**Comment by aandryashin** [comment]
Looks good to me.

**Comment by vania-pooh** [comment]
Ok, Fluxbox is 10 megabytes unpacked.
```
root@9e53ede34327:/# apt-get install fluxbox
Reading package lists... Done
Building dependency tree
Reading state information... Done
The following additional packages will be installed:
  feh libcurl3 libfribidi0 libgif7 libid3tag0 libimlib2 libjpeg-progs libjpeg9 menu xfonts-terminus
Suggested packages:
  fbpager fbdesk fbautostart menu-l10n gksu | kde-runtime | ktsuss xfonts-terminus-oblique
The following NEW packages will be installed:
  feh fluxbox libcurl3 libfribidi0 libgif7 libid3tag0 libimlib2 libjpeg-progs libjpeg9 menu xfonts-terminus
0 upgraded, 11 newly installed, 0 to remove and 73 not upgraded.
Need to get 3217 kB of archives.
After this operation, 9353 kB of additional disk space will be used.
Do you want to continue? [Y/n]
```
@bonigarcia two points:
1) To have the same behavior on both images - please add installing Fluxbox to regular (non-VNC) images (it will then automatically be installed for VNC images). Docker build files are like this: https://github.com/aerokube/selenoid-images/blob/master/selenium/opera/blink/Dockerfile.driver.tmpl
2) As you can see in `entrypoint.sh` X server is being started the last. So to avoid an error like this...
```
# fluxbox -display :99
Error: Couldn't connect to XServer:99
```
... we have to somehow try to start Fluxbox until the display is open. For example for x11vnc we add `-loop500` flag to loop until display is open. Not sure how to do this reliably for Fluxbox. But I'm afraid that without such retries depending on user's Docker speed Fluxbox will sometimes start and sometimes will not (just because X server is started after Fluxbox and this is a race condition).

**Comment by vania-pooh** [comment]
@bonigarcia btw, guys in support chat proposed to add a feature to Selenoid that will convert `/maximize` call to `setSize()` call with current screen resolution. What do you think about it? Wouldn't it be more reliable?

**Comment by bonigarcia** [comment]
@vania-pooh 

Regarding 1 (install Fluxbox in regular images), I have just committed new patches, first reverting the original patch, and then installing Fluxbox in `base_1` and `base_2`.

Regarding 2 (race condition etc), yes, you are right. In the tests I made my solution was working the 100%, but it is not the more accurate solution (start Fluxbox before the actual DISPLAY is started). In fact, I think there is an additional problem related to the VNC server. Let me explain.

In my project I am using the VNC Selenoid images. It happened to me several times that when I connect to the VNC sessions (which I always do at the beginning), sometimes the VNC session is not available. If I reload the VNC session, the second time is actually working. I have this pending for further investigation, but now I can see it is related to the wait with the `loop500` flag, which seems to be not 100% reliable (at least for my use case in which I require the VNC session right away after the container is started).

All in all, I have looked for a solution for both issues. I found it in the docker container images created by the Selenium team. See for example this [entrypoint](https://github.com/SeleniumHQ/docker-selenium/blob/master/NodeDebug/entry_point.sh). Their strategy is to launch first `xvfb-run`, then wait for it, and finally start `fluxbox` and finally `x11vnc`. I think that's the proper way, to avoid race conditions both in `fluxbox` but also in `x11vnc` (which I have experienced).

I have change all the entry points for the images implementing this strategy. I tested it and seems to work properly. Please take a look.

Regarding your last comment (the `/maximize` thing), this is not my use case. I use the VNC session to actually interact with the browsers (through noVNC), not only watching it. So I would need the browser in Docker to behave as a real browser, meaning that maximize, minimize, move the window, open preferences, etc.

File: selenium/base/base_1/Dockerfile [hunk 1]
```diff
@@ -2,7 +2,7 @@ FROM ubuntu:16.04
 
 RUN \
     apt-get update && \
-    apt-get install -y apt-transport-https ca-certificates && \
+    apt-get install -y apt-transport-https ca-certificates tzdata && \
```
Comment (vania-pooh) [inline selenium/base/base_1/Dockerfile]: No, this file will not change.

File: selenium/base/base_2/Dockerfile [hunk 2]
```diff
@@ -34,6 +34,7 @@ RUN apt-get update && apt-get install -y ttf-mscorefonts-installer \
     xfonts-encodings \
     xfonts-utils \
     flashplugin-installer \
+    fluxbox \
```
Comment (vania-pooh) [inline selenium/base/base_2/Dockerfile]: Adding fluxbox here means I have to recreate all browser images as they are all using this layer.

File: selenium/firefox/selenium/entrypoint.sh [hunk 3]
```diff
@@ -1,4 +1,19 @@
 #!/bin/bash
 SCREEN_RESOLUTION=${SCREEN_RESOLUTION:-"1920x1080x24"}
 DISPLAY=99
-/usr/bin/xvfb-run -l -n "$DISPLAY" -s "-ac -screen 0 $SCREEN_RESOLUTION -noreset -listen tcp" /usr/bin/java -Xmx256m -Djava.security.egd=file:/dev/./urandom -jar /usr/share/selenium/selenium-server-standalone.jar -port 4444 -browserTimeout 120
\ No newline at end of file
+/usr/bin/xvfb-run -l -n "$DISPLAY" -s "-ac -screen 0 $SCREEN_RESOLUTION -noreset -listen tcp" /usr/bin/java -Xmx256m -Djava.security.egd=file:/dev/./urandom -jar /usr/share/selenium/selenium-server-standalone.jar -port 4444 -browserTimeout 120 &
+XVFB_PID=$!
+
+for i in $(seq 1 10)
```
Comment (vania-pooh) [inline selenium/firefox/selenium/entrypoint.sh]: How about using `until` here? E.g. like this: https://github.com/aerokube/selenoid-images/blob/master/selenium/video/entrypoint.sh#L9

File: selenium/base/base_1/Dockerfile (hunk 1 above)
Comment (bonigarcia) [reply selenium/base/base_1/Dockerfile]: This package is required if `base_1` is intended to be installed:

```
RUN \
    echo ttf-mscorefonts-installer msttcorefonts/accepted-mscorefonts-eula select true | debconf-set-selections && \
    echo 'UTC' | tee /etc/timezone && \
    dpkg-reconfigure -f noninteractive tzdata && \
    echo "gtk-cursor-blink=0" > /root/.gtkrc-2.0
```
In any case if you prefer I can drop it, but `base_1` would useless again.

File: selenium/base/base_2/Dockerfile (hunk 2 above)
Comment (bonigarcia) [reply selenium/base/base_2/Dockerfile]: The other alternative is to install fluxbox each by one in the browser images, but I think if you want to be used in all containers, it is a better to keep this package in the base.

File: selenium/firefox/selenium/entrypoint.sh (hunk 3 above)
Comment (bonigarcia) [reply selenium/firefox/selenium/entrypoint.sh]: Yes, good idea. I have added a new patch with that.

File: selenium/chrome/entrypoint.sh [hunk 4]
```diff
@@ -1,4 +1,19 @@
...
 SCREEN_RESOLUTION=${SCREEN_RESOLUTION:-"1920x1080x24"}
 DISPLAY=99
-/usr/bin/xvfb-run -l -n "$DISPLAY" -s "-ac -screen 0 $SCREEN_RESOLUTION -noreset -listen tcp" /usr/bin/chromedriver --port=4444 --whitelisted-ips='' --verbose
\ No newline at end of file
+/usr/bin/xvfb-run -l -n "$DISPLAY" -s "-ac -screen 0 $SCREEN_RESOLUTION -noreset -listen tcp" /usr/bin/chromedriver --port=4444 --whitelisted-ips='' --verbose &
+XVFB_PID=$!
+
+for i in $(seq 1 10)
+do
+  xdpyinfo -display $DISPLAY >/dev/null 2>&1
```
[the remark this answers is not in the document]
Comment (bonigarcia) [reply selenium/chrome/entrypoint.sh]: Yes, my mistake. I have added the Ubuntu package required to install this binary (`x11-utils`).

**Comment by bonigarcia** [comment]
@vania-pooh Unfortunately the `/maximize`, even much more simpler, it is not enough for me. Regarding the other requests, I have added some comments and committed a new patch.

**Comment by vania-pooh** [comment]
@bonigarcia what else is covered by Fluxbox in addition to `/maximize`?

**Comment by bonigarcia** [comment]
I already explained it. Without a windows manager you cannot control the control the browsers, meaning that you cannot maximize, minimize, move, or open preferences. I use noVNC to interact with the VNC browser session (it's a bidirectional communication based on WebSocket). It's not only watching the browser with a given size.

**Comment by vania-pooh** [comment]
@bonigarcia sorry for the delay. I reviewed again in detail your PR. Everything seems to be ok. I would request one more change. Please move installing `fluxbox` and `x11-utils` from base image to the webdriver layer before actually copying the binary. For example insert something like this [here](https://github.com/aerokube/selenoid-images/blob/master/selenium/chrome/Dockerfile.driver.tmpl#L3):
```
RUN \
        apt-get update && \
        apt-get -y  install fluxbox x11-utils && \
        apt-get clean && \
        rm -Rf /tmp/* && rm -Rf /var/lib/apt/lists/*
```
I understand that this is about copy-pasting and I will move these packages to `base_3.0` image when needed to create it. Currently this would be easier and faster to use layers with browsers and install additional 10 megabytes on top.

**Comment by vania-pooh** [comment]
Thank you for contribution! Will rebuild images during next weeks.

**Comment by bonigarcia** [comment]
@vania-pooh Thanks to you! Keep up the great work with Selenoid!

**Comment by vania-pooh** [comment]
@bonigarcia just tried to create an image for Chrome 65 using new Dockerfiles. Window maximization does not seem to work (getting an exception when doing `maximize` call). As usually these images are available as: `selenoid/chrome:65.0` and `selenoid/vnc:chrome_65.0`. Any ideas what's the reason?

**Comment by bonigarcia** [comment]
@vania-pooh Yes, it seems that the Flux windows manager is missing in the new image for Chrome 65.

We can check it starting the Chrome VNC container as standalone and checking the traces:

```
$ docker run --rm --name chrome --cap-add=SYS_ADMIN -p 4444:4444 -p 5900:5900 selenoid/vnc:chrome_65.0

 --- x11vnc loop: 1 ---

Starting ChromeDriver 2.36.540471 (9c759b81a907e70363c6312294d30b6ccccc2752) on port 4444
All remote connections are allowed. Use a whitelist instead!
 --- x11vnc loop: waiting for: 34

PORT=5900
```
As we can see in the latest [entrypoint.sh](https://github.com/aerokube/selenoid-images/blob/master/selenium/vnc/chrome/entrypoint.sh), the VNC Chrome container is supposed to wait for `xvfb-run` before starting `fluxbox`. Notice that the traces `Waiting xvfb...` and the traces for `fluxbox` are completely missing.

Maybe the Chrome 65 image has been generated without this PR?

**Comment by vania-pooh** [comment]
Ok, let me investigate more.

**Comment by vania-pooh** [comment]
@bonigarcia just with `driver.manage().window().maximize();` (Java client).

**Pull request merged**